moto/utilities/id_generator.py

```diff
@@ -0,0 +1,118 @@
+import abc
+import threading
+from typing import Any, Callable, Dict, List, Union
+
+from moto.moto_api._internal import mock_random
+
+ExistingIds = Union[List[str], None]
+
+
+class ResourceIdentifier(abc.ABC):
+    """
+    Base class for resource identifiers. When implementing a new resource, it is important to set
+    the service and resource as they will be used to create the unique identifier for that resource.
+
+    It is recommended to implement the `generate` method using functions decorated with `@moto_id`.
+    This will ensure that your resource can be assigned a custom id.
+    """
+
+    service: str
+    resource: str
+
+    def __init__(self, account_id: str, region: str, name: str):
+        self.account_id = account_id
+        self.region = region
+        self.name = name or ""
+
+    @abc.abstractmethod
+    def generate(self, existing_ids: ExistingIds = None) -> str: ...
+
+    """ If `existing_ids` is provided, we will not return a custom id if it is already on the list"""
+
+    @property
+    def unique_identifier(self) -> str:
+        return ".".join(
+            [self.account_id, self.region, self.service, self.resource, self.name]
+        )
+
+
+class MotoIdManager:
+    """class to manage custom ids. Do not create instance and instead
+    use the `id_manager` instance created below."""
+
+    _custom_ids: Dict[str, str]
+    _id_sources: List[Callable[[ResourceIdentifier], Union[str, None]]]
+
+    _lock: threading.RLock
+
+    def __init__(self) -> None:
+        self._custom_ids = {}
+        self._lock = threading.RLock()
+        self._id_sources = []
+
+        self.add_id_source(self.get_custom_id)
```

Comment:
What's the rationale behind having multiple ID sources?

Reply:
Having a deterministic fallback would make sense here, considering `moto_id_generator` is the only place where LS can tweak the behaviour.